### PR TITLE
Add API for external Buffer component iteration

### PR DESCRIPTION
**Pull request by chrisvest**
Motivation:
Accessing the internal ByteBuffer instances, or the native addresses, with the `Buffer.forEachReadable/Writable` methods have turned out to be rather troublesome.
The trouble specifically comes from the API being designed around _internal_ iteration.
While internal iteration is excellent at masking whether a buffer is composite or not, the loss of local mutable context (because of the use of lambdas) makes it rather combersome to use.

Modification:
This adds an API based on external iteration of the components.
There are a number of requirements pulling on this API and causing it to look the way it does.
First, we really want to ensure that the parent `Buffer` instance is strongly referenced and live throughout the entire iteration.
With internal iteration, we can place a `reachabilityFence` after the iteration.
With external iteration, we instead rely on the `ComponentIterator` being `AutoCloseable` and place a reachability fence in the close method.
Second, we wish to prevent the components from leaking beyond the iteration.
This isn't entirely possible to prevent, for either internal or external iteration.
The use of intersection types discourages assigning the iterator or the components to fields, though they don't prevent it.
It does, however, prevent correctly typed local variables from existing beyond their intended scope.

Result:
We have a API for external buffer component iteration.

### CODE REVIEW

File: buffer/src/main/java/io/netty5/buffer/api/Buffer.java [hunk 1]
```diff
@@ -983,6 +984,32 @@ default Buffer split() {
...
+     * The typical code pattern for using this API looks like the following:
+     * <pre>{@code
+     *      try (var iteration = buffer.forEachReadable()) {
+     *          for (var c = iteration.first(); c != null; c = c.next()) {
+     *              ByteBuffer componentBuffer = c.readableBuffer();
+     *              // ...
+     *          }
+     *      }
+     * }</pre>
+     * Note the use of the {@code var} keyword for local variables, which are required for correctly expressing the
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty5/buffer/api/Buffer.java:1002]: see above the question

File: buffer/src/main/java/io/netty5/buffer/api/ComponentIterator.java [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty5.buffer.api;
+
+import io.netty5.buffer.api.ComponentIterator.Next;
+
+public interface ComponentIterator<T extends Next> extends AutoCloseable {
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty5/buffer/api/ComponentIterator.java:20]: I wonder why we cant just use `MyIterator<T> extends, Iterator<T>, AutoCloseable` ? Is it because you want to ensure the user always close ?

Comment (chrisvest) [reply buffer/src/main/java/io/netty5/buffer/api/ComponentIterator.java:20]: I'm trying to avoid allocations for single-component buffers, but maybe that's still possible with the JDK iterator interface.

File: buffer/src/main/java/io/netty5/buffer/api/Buffer.java [hunk 3]
```diff
@@ -983,6 +984,32 @@ default Buffer split() {
...
+    /**
+     * Create a {@linkplain ComponentIterator component iterator} for all readable components in this buffer.
+     * <p>
+     * Unlike the {@link #forEachReadable(int, ReadableComponentProcessor)} method, this API permits <em>external</em>
+     * iteration of the components, while at the same time protecting the life-cycle of the buffer.
+     * <p>
+     * The typical code pattern for using this API looks like the following:
+     * <pre>{@code
+     *      try (var iteration = buffer.forEachReadable()) {
+     *          for (var c = iteration.first(); c != null; c = c.next()) {
```
[the remark this answers is not in the document]
Comment (chrisvest) [reply buffer/src/main/java/io/netty5/buffer/api/Buffer.java:996]: Then you have to choose one of the intersecting interface, and you loose either the `ReadableComponent` methods, or the `next()` method from `ComponentIterator.Next`.

[the remark this answers is not in the document]
Comment (normanmaurer) [reply buffer/src/main/java/io/netty5/buffer/api/Buffer.java:996]: got it...

File: buffer/src/main/java/io/netty5/buffer/api/ComponentIterator.java (hunk 2 above)
Comment (franz1981) [reply buffer/src/main/java/io/netty5/buffer/api/ComponentIterator.java:20]: just if the iterator is cached and reused on close I think

**Comment by normanmaurer** [comment]
@chrisvest I like the abstraction... Let's do it

File: buffer/src/main/java/io/netty5/buffer/api/ComponentIterator.java (hunk 2 above)
Comment (franz1981) [reply buffer/src/main/java/io/netty5/buffer/api/ComponentIterator.java:20]: Not sure, depends if between allocation and iterator close there are operations which the JVM cannot figure it out what will happen next - meaning that it cannot mark the object as won't escape.
I can write a jmh bench to check it out btw, If it can help and see if it allocate anything

Comment (normanmaurer) [inline buffer/src/main/java/io/netty5/buffer/api/ComponentIterator.java:20]: Add javadocs

File: buffer/src/main/java/io/netty5/buffer/api/internal/SingleComponentIterator.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import io.netty5.buffer.api.ComponentIterator.Next;
+import io.netty5.buffer.api.Resource;
+
+import java.lang.ref.Reference;
+
+public final class SingleComponentIterator<T extends Next> implements ComponentIterator<T> {
+    private final T singleComponent;
+
+    @SuppressWarnings("unchecked")
+    public SingleComponentIterator(Object singleComponent) {
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty5/buffer/api/internal/SingleComponentIterator.java:28]: Why you cant use `T extends Next` ?

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: buffer/src/main/java/io/netty5/buffer/api/internal/SingleComponentIterator.java (hunk 4 above)
Comment (chrisvest) [reply buffer/src/main/java/io/netty5/buffer/api/internal/SingleComponentIterator.java:28]: Because the generic type at the call site would actually have stricter bounds, so we would need to cast there. Effectively, this moves the cast into the constructor, which I think makes the code look less busy, but we can do either way.

File: buffer/src/main/java/io/netty5/buffer/api/ComponentIterator.java (hunk 2 above)
Comment (chrisvest) [reply buffer/src/main/java/io/netty5/buffer/api/ComponentIterator.java:20]: There are multiple things that need to have their javadocs expanded, and also tests needs to be added. I'm working on that.

File: buffer/src/main/java/io/netty5/buffer/api/internal/SingleComponentIterator.java (hunk 4 above)
Comment (normanmaurer) [reply buffer/src/main/java/io/netty5/buffer/api/internal/SingleComponentIterator.java:28]: its fine ... just wanted to know if I am missing anything

**Comment by chrisvest** [comment]
@normanmaurer Comments addressed. Please take a look.

File: buffer/src/main/java/io/netty5/buffer/api/DefaultCompositeBuffer.java [hunk 5]
```diff
@@ -1951,4 +1965,183 @@ public boolean process(int index, WritableComponent component) {
...
+        @Override
+        public ByteCursor openCursor() {
+            return ((ReadableComponent) currentComponent).openCursor();
+        }
+
+        @Override
+        public void skipReadable(int byteCount) {
+            if (byteCount < 0) {
+                throw new IllegalArgumentException("Byte count cannot be negative: " + byteCount);
+            }
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty5/buffer/api/DefaultCompositeBuffer.java:2087]: nit: I guess we could remove this check above as the `ReadableComponent` should check itself anyway ?

File: buffer/src/main/java/io/netty5/buffer/api/DefaultCompositeBuffer.java [hunk 6]
```diff
@@ -1951,4 +1965,183 @@ public boolean process(int index, WritableComponent component) {
...
+        public ByteBuffer writableBuffer() {
+            return ((WritableComponent) currentComponent).writableBuffer();
+        }
+
+        @Override
+        public void skipWritable(int byteCount) {
+            if (byteCount < 0) {
+                throw new IllegalArgumentException("Byte count cannot be negative: " + byteCount);
+            }
+            ((WritableComponent) currentComponent).skipWritable(byteCount);
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty5/buffer/api/DefaultCompositeBuffer.java:2133]: same comment as above

File: buffer/src/main/java/io/netty5/buffer/api/adaptor/ByteBufBuffer.java [hunk 7]
```diff
@@ -613,6 +709,9 @@ public int getUnsignedByte(int roff) {
 
     @Override
     public Buffer writeByte(byte value) {
+        if (readOnly()) {
+            throw bufferIsReadOnly(this);
+        }
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty5/buffer/api/adaptor/ByteBufBuffer.java:714]: should we better do these changes in another PR and add the relevant tests there ? This seems completely unrelated to this change

File: buffer/src/main/java/io/netty5/buffer/api/internal/SingleComponentIterator.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+import java.lang.ref.Reference;
+
+public final class SingleComponentIterator<T extends Next> implements ComponentIterator<T> {
+    private final Resource<?> lifecycle;
+    private final T singleComponent;
+
+    @SuppressWarnings("unchecked")
+    public SingleComponentIterator(Resource<?> lifecycle, Object singleComponent) {
+        this.lifecycle = lifecycle;
+        this.singleComponent = (T) singleComponent;
```
Comment (normanmaurer) [inline buffer/src/main/java/io/netty5/buffer/api/internal/SingleComponentIterator.java:31]: add null checks

**Review by normanmaurer — CHANGES_REQUESTED** [review]
almost there

File: buffer/src/main/java/io/netty5/buffer/api/adaptor/ByteBufBuffer.java (hunk 7 above)
Comment (chrisvest) [reply buffer/src/main/java/io/netty5/buffer/api/adaptor/ByteBufBuffer.java:714]: https://github.com/netty/netty/pull/12355

Comment (chrisvest) [reply buffer/src/main/java/io/netty5/buffer/api/adaptor/ByteBufBuffer.java:714]: When that one is merged, I can rebase this PR on top. The component iteration tests run into this problem, but in a more obscure way, so it needs to be fixed before this PR can go in.

**Comment by chrisvest** [comment]
@normanmaurer Comments addressed and rebased

**Review by normanmaurer — APPROVED** [review]
